libraries/opensk/src/env/test/rust_crypto.rs

```diff
@@ -0,0 +1,254 @@
+// Copyright 2023 Google LLC
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//      http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+//! This cryptography implementation is an alternative for our own library.
+//!
+//! You can use it with the `rust_crypto` feature. An example call to cargo test is in
+//! `run_desktop_tests.sh`. It is currently impossible to use it with our version of TockOS due to
+//! a compiler version imcompatibility.
+//!
+//! If you want to use OpenSK outside of Tock v1, maybe this is useful for you though!
+
+use crate::api::crypto::{ecdh, ecdsa, Crypto, EC_FIELD_BYTE_SIZE, EC_SIGNATURE_SIZE};
+use core::convert::TryFrom;
+use p256::ecdh::EphemeralSecret;
+use p256::ecdsa::signature::{SignatureEncoding, Signer, Verifier};
+use p256::ecdsa::{SigningKey, VerifyingKey};
+use p256::elliptic_curve::sec1::ToEncodedPoint;
+use rand_core::OsRng;
+use rng256::Rng256;
+
+pub struct TestCrypto;
+pub struct TestEcdh;
+pub struct TestEcdsa;
+
+impl Crypto for TestCrypto {
+    type Ecdh = TestEcdh;
+    type Ecdsa = TestEcdsa;
+}
+
+impl ecdh::Ecdh for TestEcdh {
+    type SecretKey = TestEcdhSecretKey;
+    type PublicKey = TestEcdhPublicKey;
+    type SharedSecret = TestEcdhSharedSecret;
+}
+
+pub struct TestEcdhSecretKey {
```

Comment:
For both implementations (`crypto.rs` and `rust_crypto.rs`), why use a wrapper type? Can't we just do the following?

```rust
pub type TestEcdhSecretKey = EphemeralSecret;
```

Reply:
I think it will make the implementation much more obvious/readable by removing the boilerplate.

Reply:
But regardless, I see how it can be problematic to have 2 traits implemented for the same type with the same method names. Fine with the current version.

Reply:
Nice, so we don't have to be as careful when calling. FYI your other call idea throws this:

```
error[E0277]: the trait bound `TestRng256: CryptoRng` is not satisfied
    --> src/ctap/mod.rs:3509:59
     |
3509 |         let key_agreement_key = EcdhSk::<TestEnv>::random(env.rng());
     |                                 ------------------------- ^^^^^^^^^ the trait `CryptoRng` is not implemented for `TestRng256`
```

So looks like it is still confused.

Reply:
Yes this means it thought it was the other trait...